File: latte/src/main/java/gg/beemo/latte/broker/LocalConnection.kt
package gg.beemo.latte.broker

class LocalConnection(
    override val serviceName: String,
    override val instanceId: String,
) : BrokerConnection() {

    override val supportsTopicHotSwap = true

    override suspend fun abstractSend(
        topic: String,
        key: String,
        value: String,
        headers: BrokerMessageHeaders,
    ): MessageId {
        if (shouldDispatchExternallyAfterShortCircuit(topic, key, value, headers)) {
            val isForExternalService =
                headers.targetServices.isNotEmpty() && headers.targetServices != setOf(serviceName)
            val isForExternalInstance =
                headers.targetInstances.isNotEmpty() && headers.targetInstances != setOf(instanceId)
            if (isForExternalService || isForExternalInstance) {
                throw IllegalArgumentException(
                    "Attempting to send message to other services/instances " +
                            "(services=${headers.targetServices}; instances=${headers.targetInstances}) " +
                            "in a LocalConnection"
                )
            }
        }

        return headers.messageId
    }

    override suspend fun start() {
        // Nothing to start :)
    }

    override fun createTopic(topic: String) {
        // noop
    }

    override fun removeTopic(topic: String) {
        // noop
    }

}

File: water/src/broker/LocalConnection.ts
import type { MessageId } from "./BrokerConnection.js";
import { BrokerConnection } from "./BrokerConnection.js";
import type { BrokerMessageHeaders } from "./BrokerMessageHeaders.js";

export class LocalConnection extends BrokerConnection {

    public override readonly supportsTopicHotSwap: boolean = true;
    public override readonly deferInitialTopicCreation: boolean = false;

    public constructor(public override readonly serviceName: string, public override readonly instanceId: string) {
        super();
    }

    public override async abstractSend(topic: string, key: string, value: string, headers: BrokerMessageHeaders): Promise<MessageId> {
        if (this.shouldDispatchExternallyAfterShortCircuit(topic, key, value, headers)) {
            const isForExternalService = headers.targetServices.size > 0 && [...headers.targetServices][0] !== this.serviceName;
            const isForExternalInstance = headers.targetInstances.size > 0 && [...headers.targetInstances][0] !== this.instanceId;
            if (isForExternalService || isForExternalInstance) {
                throw new Error("Attempting to send message to other services/instances in a LocalConnection " +
                    `(services=${[...headers.targetServices]}; instances=${[...headers.targetInstances]})`);
            }
        }
        return headers.messageId;
    }

    public override async abstractStart(): Promise<void> {
        // Nothing to start :)
    }

    public override createTopic(topic: string): void {
        // noop
    }

    public override removeTopic(topic: string): void {
        // noop
    }

}
